{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7359d202-bf95-4dea-be9b-75d0bbb15a32",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Introduction\n",
    "\n",
    "## Download and get data\n",
    "\n",
    "Go to: `~/appladmix/notebooks/2_data/data/`\n",
    "\n",
    "`wget https://ucloud.univie.ac.at/index.php/s/sjVDEgg2KDvI9u8/download`\n",
    "\n",
    "`tar -zxvf download`\n",
    "\n",
    "\n",
    "### The 1000 Genomes Project\n",
    "\n",
    "The 1000 Genomes Project is a commonly-used reference data resources in human genetic studies.\n",
    "\n",
    "In the phase 3 data, there are 2,504 individuals from 26 populations ([Sudmant et al. 2015](https://doi.org/10.1038/nature15394)).\n",
    "\n",
    "<img src=\"https://upload.wikimedia.org/wikipedia/commons/thumb/e/ec/1000_Genomes_Project.svg/1280px-1000_Genomes_Project.svg.png?1657572184698\" alt=\"1KG\" width=\"1000\"/>\n",
    "\n",
    "**Figure 2 Locations of population samples of 1000 Genomes Project.** Each circle represents the number of sequences in the final release. This figure is from https://www.wikiwand.com/en/1000_Genomes_Project#/Human_genome_samples.\n",
    "\n",
    "In this course, we usually use genotype data from the YRI and CEU populations as examples.\n",
    "\n",
    "Here,\n",
    "- YRI: Yoruba in Ibadan, Nigeria.\n",
    "- CEU: Utah Residents (CEPH) with Northern and Western European ancestry.\n",
    "\n",
    "More information can be found in their [website](https://www.internationalgenome.org/).\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b6b36ff5",
   "metadata": {},
   "source": [
    "# Handling data in a VCF file\n",
    "\n",
    "VCF files have columns (individuals) and rows (genotypes). We can add columns with `bcftools merge`, and select columns with *e.g.* `bcftools view -s`. Likewise, we can add rows with `bcftools concat` and select rows with *e.g.* `bcftools view -r`.\n",
    "\n",
    "**Concatenating:**\n",
    "\n",
    "![image.png](attachment:image.png)\n",
    "\n",
    "**Merging:**\n",
    "\n",
    "![image-2.png](attachment:image-2.png)\n",
    "\n",
    "So let's first try to get a subset of rows:\n",
    "\n",
    "```\n",
    "bcftools view -H -r 21:10000000-15000000 chr21.YRI.CEU.vcf.gz | less\n",
    "```\n",
    "\n",
    "## Subsetting\n",
    "\n",
    "Here, we mainly want to do subsetting to get good data. If we want to extract data from some individuals, for example, samples from the same population, we can use the argument `-S` with a file containing names of samples you want to extract. \n",
    "\n",
    "Here, we provide two example files `YRI.list` and `CEU.list` with samples from the YRI and CEU populations.\n",
    "\n",
    "An example from `YRI.list` is below. Each line contains a sample name.\n",
    "\n",
    "```\n",
    "NA18486\n",
    "NA18488\n",
    "NA18489\n",
    "NA18498\n",
    "NA18499\n",
    "```\n",
    "\n",
    "Then we can extract genetic variants from the YRI and CEU populations with the following commands, respectively.\n",
    "\n",
    "```\n",
    "bcftools view chr21.YRI.CEU.vcf.gz -S YRI.list | bgzip -c > chr21.YRI.vcf.gz\n",
    "bcftools view chr21.YRI.CEU.vcf.gz -S CEU.list | bgzip -c > chr21.CEU.vcf.gz\n",
    "```\n",
    "\n",
    "Let's have a look at the files!\n",
    "\n",
    "\n",
    "## Filtering data\n",
    "\n",
    "When analysing data, we usually want to use variants that can meet some conditions, for example, variants with good quality. Then we can use several arguments to filter data in VCF files.\n",
    "\n",
    "For example, the following commands extract biallelic single nucleotide polymorphisms (SNPs) that passed quality checks from the YRI and CEU populations.\n",
    "\n",
    "```\n",
    "bcftools view chr21.YRI.vcf.gz -m 2 -M 2 -v snps | bgzip -c > chr21.YRI.biallelic.snps.vcf.gz\n",
    "bcftools view chr21.CEU.vcf.gz -m 2 -M 2 -v snps | bgzip -c > chr21.CEU.biallelic.snps.vcf.gz\n",
    "```\n",
    "\n",
    "What is the meaning of the arguments `-m`, `-M`, and `-v`? Ask bcftools directly: `bcftools view -h` \n",
    "\n",
    "Or use the [manual](https://samtools.github.io/bcftools/bcftools.html#view)!\n",
    "\n",
    "One can do much more filtering with `bcftools`, but this should suffice for now.\n",
    "\n",
    "\n",
    "## Adding data\n",
    "\n",
    "Very good. I assume we are all somewhat interested in ***admixture***. Ok, how about using Neanderthals and Denisovans as possible source populations? \n",
    "\n",
    "Let's download the vcf files:\n",
    "\n",
    "```\n",
    "wget https://ucloud.univie.ac.at/index.php/s/enPq6XjDqkaeXRW/download\n",
    "\n",
    "tar -zxvf download.1\n",
    "\n",
    "bcftools index altai.filt.vcf.gz\n",
    "bcftools index denisovan.filt.vcf.gz\n",
    "```\n",
    "\n",
    "\n",
    "And perform a merging operation:\n",
    "\n",
    "```\n",
    "bcftools merge -o chr21.merged.vcf.gz -g /home/local/ANTHROPOLOGY/kuhlwilmm83/refgen/hg19/hg19.p13.plusMT.no_alt_analysis_set.fa.gz chr21.YRI.CEU.vcf.gz altai.filt.vcf.gz denisovan.filt.vcf.gz\n",
    "```\n",
    "\n",
    "Let's have a look at the file!\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc419d21",
   "metadata": {},
   "source": [
    "## Data analysis\n",
    "\n",
    "More important is to get some idea what we have there. That is, for example, what is the diversity of these populations?\n",
    "\n",
    "Let's start with heterozygosity (how many heterozygous sites are there in this chromosome?). It is very easy to obtain, again with bcftools:\n",
    "\n",
    "```\n",
    "bcftools stats -s -  chr21.merged.vcf.gz | grep \"^PSC\" -B 1 > hets.txt\n",
    "```\n",
    "\n",
    "If you inspect the file, you will see that there is one line per sample, with several columns containing the statistics. Column 6 is the number of Hets.\n",
    "\n",
    "Maybe this is easier in `R`?\n",
    "\n",
    "We can:\n",
    "* run R from the command line\n",
    "* read in these summary tables\n",
    "* plot a histogram\n",
    "* calculate the mean for CEU and YRI separately, compared to the archaics\n",
    "* perform a statistical test of CEU and YRI are different\n",
    "* There is ~30 million bases for which we can have data on chr21, often heterozygosity is expressed als hets per 1,000bp. We can present the values in such a way.\n",
    "\n",
    "```\n",
    "R --vanilla\n",
    "\n",
    "hets<-read.table(\"hets.txt\", sep=\"\\t\",header=T,comment.char=\"\")\n",
    "\n",
    "png(\"hets_histogram.png\",600,600)\n",
    "hist(hets[,6],breaks=20)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "ceu<-unlist(read.table(\"CEU.list\", sep=\"\\t\",header=F))\n",
    "yri<-unlist(read.table(\"YRI.list\", sep=\"\\t\",header=F))\n",
    "\n",
    "wilcox.test(hets[which(hets[,3]%in%ceu),6],hets[which(hets[,3]%in%yri),6])\n",
    "\n",
    "genom=30000000\n",
    "\n",
    "mean(hets[which(hets[,3]%in%ceu),6])/genom*1000\n",
    "mean(hets[which(hets[,3]%in%yri),6])/genom*1000\n",
    "hets[which(hets[,3]%in%c(\"AltaiNea\",\"DenisovaPinky\")),6]/genom*1000\n",
    "\n",
    "```\n",
    "\n",
    "## Genetic diversity\n",
    "\n",
    "The number of heterozygous positions reflects an evolutionarily important feature, as it is a very direct metric of genetic diversity.\n",
    "\n",
    "* We can interpret the numbers above in the context of evolution.\n",
    "* Which evolutionary forces can influence the observed heterozygosity?\n",
    "* Which technical problems can influence it?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71d95c7f",
   "metadata": {},
   "source": [
    "# Now, let's go for the first challenge!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
