## Applications of Admixture Genomics (UE UniVienna).

This repo contains the materials for practical sessions, and will be updated accordingly. The current implementation is based on the LISC, where JupyterLab and RStudio are provided. User accounts are bound to enrolment in the course, and access is associated to this course. However, all notebooks are freely available, and the data is in principle public as well.

It is expected to have basic knowledge on using bash and a unix-based command line environment, with a reminder in the "Session 0".

This part can actually be done anywhere using [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/admixVIE/appladmix/HEAD). This will create an online virtual machine containing a basic setting for bash and R. Specific programs for later parts of the course will not run, but it can help to become familiar (again) with this setting.



At the moment, there are the following modules:
- Session 0 (before the course): Basics in bash ("0_basics").
- Session 1: Handling of sequencing and genotype data ("1_data")
- Session 2: Exploratory population genetic analysis; with Challenge ("2_popgen") 
- Session 3: Admixtools, f/D-statistics ("3_admixtools")
- Session 4: Introgression detection with sstar; with Challenge ("4_introgression")



Planned modules are:
- Further population genetics with NGStools, plink
- Handling and exploring ancient DNA sequencing data
- Introgression detection with other tools (in development)
- Simulating data

