{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "05179d61",
   "metadata": {},
   "source": [
    "# Challenge 2 solution\n",
    "\n",
    "Pre-challenge:\n",
    "```\n",
    "bcftools view -a -s ^\"DenisovaPinky,AltaiNea\" testset.filtered.vcf.gz | bcftools view -e 'GT=\"./.\"' -m 2 -M 2 | bgzip > testset.filtered.nomis.vcf.gz\n",
    "```\n",
    "\n",
    "\n",
    "## A)\n",
    "\n",
    "Install gdc (and, as in the previous lecture, the admixtools package):\n",
    "\n",
    "```\n",
    "cd ~/\n",
    "git clone https://github.com/mathii/gdc.git\n",
    "\n",
    "R --vanilla\n",
    "install.packages(c(\"igraph\",\"dplyr\",\"plotly\"),repos=\"https://cloud.r-project.org\",upgrade=\"never\")\n",
    "devtools::install_github(\"uqrmaie1/admixtools\",repos=\"https://cloud.r-project.org\",upgrade=\"never\")\n",
    "library(\"admixtools\")\n",
    "```\n",
    "\n",
    "Get data to eigenstrat:\n",
    "```\n",
    "python2 ~/gdc/vcf2eigenstrat.py -v testset.filtered.vcf.gz -o testset -i testset_1.ind\n",
    "```\n",
    "\n",
    "Analyse with admixtools:\n",
    "\n",
    "```\n",
    "R --vanilla\n",
    "library(admixtools)\n",
    "library(dplyr)\n",
    "options(scipen=100)\n",
    "\n",
    "extract_f2(pref=\"testset\",\n",
    "    outdir=\"genos\",blgsize=500000,\n",
    "    overwrite=T,maxmem=500\n",
    ")\n",
    "\n",
    "f2_blocks = f2_from_precomp(\"genos\")\n",
    "f4_table1<-f4(f2_blocks,pop1=\"YRI\", pop2=\"IBS\",pop3=\"NEA\", pop4=\"DEN\")\n",
    "```\n",
    "\n",
    "IBS has more shared alleles with NEA than expected. This is due to gene flow from Neandertals to non-Africans. The z-score is low because this is only one chromosome.\n",
    "\n",
    "```\n",
    "pop1  pop2  pop3  pop4       est       se     z      p\n",
    "  <chr> <chr> <chr> <chr>    <dbl>    <dbl> <dbl>  <dbl>\n",
    "1 YRI   IBS   NEA   DEN   -0.00202 0.000905 -2.24 0.0254\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40bcc1f7",
   "metadata": {},
   "source": [
    "## B)\n",
    "\n",
    "Calculate matchrate to Denisovan:\n",
    "```\n",
    "sstar matchrate --vcf testset.filtered.nomis.witharch.vcf.gz --ref ref.ind.list --tgt tgt.ind.list --src deni.list --score testset.threshold.results --mapped-region testset.tracts.bed --output deni.match.rate.results\n",
    "```\n",
    "\n",
    "Get Denisovan-like fragments regardless of Neanderthals:\n",
    "```\n",
    "sstar tract --threshold testset.threshold.results --output-prefix deni.final --match-rate deni.match.rate.results\n",
    "```\n",
    "\n",
    "Count the number of fragments:\n",
    "```\n",
    "R --vanilla # or RStudio\n",
    "\n",
    "setwd(\"/home/jovyan/notebooks/introgression/data\")\n",
    "matchtab<-read.table(\"deni.final.bed\",sep=\"\\t\",header=F)\n",
    "as.data.frame(table(matchtab[which(matchtab[,5]>0.33),4]))\n",
    "```\n",
    "\n",
    "Do a direct comparison what fits best, following the sstar tutorial:\n",
    "\n",
    "```\n",
    "sstar tract --threshold testset.threshold.results --output-prefix both.final --match-rate altai.match.rate.results deni.match.rate.results\n",
    "```\n",
    "\n",
    "Count the number of fragments:\n",
    "```\n",
    "R --vanilla # or RStudio\n",
    "\n",
    "setwd(\"/home/jovyan/notebooks/introgression/data\")\n",
    "matchtab<-read.table(\"both.final.src2.bed\",sep=\"\\t\",header=F)\n",
    "as.data.frame(table(matchtab[which(matchtab[,5]>0.33),4]))\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98eb2e7f",
   "metadata": {},
   "source": [
    "## C)\n",
    "\n",
    "The following is a rather efficient solution (not even the most efficient one). As always, there are different ways of reaching the same goal.\n",
    "\n",
    "Get values of cumulative length per individual on the command line (one could also do the same in R, but for practical purposes we stick to the methods from the course). The easiest is a `while` loop, but it can be done one by one with copy-pasting of code.\n",
    "\n",
    "```\n",
    "while read ind in\n",
    "do\n",
    "grep \"$ind\" both.final.src2.bed | bedtools sort | bedtools merge > $ind.deni.bed\n",
    "grep \"$ind\" both.final.src1.bed | bedtools sort | bedtools merge > $ind.altai.bed\n",
    "done < tgt.ind.list\n",
    "```\n",
    "\n",
    "Otherwise, just do `ind=HG02223` and the same two lines with grep, repeat with the next `ind` etc.\n",
    "\n",
    "\n",
    "Load tables into R. The most straightforward is a `for` loop in R, repeat the same action and collect the output:\n",
    "\n",
    "```\n",
    "R --vanilla # or RStudio\n",
    "setwd(\"/home/jovyan/notebooks/introgression/data\")\n",
    "indlist<-unlist(read.table(\"tgt.ind.list\",sep=\"\\t\",header=F))\n",
    "\n",
    "lengsA<-c()\n",
    "numsA<-c()\n",
    "lengsD<-c()\n",
    "numsD<-c()\n",
    "for (j in (1:length(indlist))) { \n",
    "    matchtab1<-read.table(paste(indlist[j],\".altai.bed\",sep=\"\"),sep=\"\\t\",header=F)\n",
    "    matchtab2<-read.table(paste(indlist[j],\".deni.bed\",sep=\"\"),sep=\"\\t\",header=F)\n",
    "    lengsA[j]<-sum(matchtab1[,3]-matchtab1[,2])\n",
    "    lengsD[j]<-sum(matchtab2[,3]-matchtab2[,2])\n",
    "    numsA[j]<-nrow(matchtab1)\n",
    "    numsD[j]<-nrow(matchtab2)\n",
    "    }\n",
    "```\n",
    "\n",
    "Now, create simple barplots:\n",
    "\n",
    "```\n",
    "nicecol <- colorRampPalette(c(\"orange\", \"blue\"))(10)\n",
    "barplot(lengsA,beside=T,names.arg=indlist,col=c(nicecol),las=2)\n",
    "barplot(lengsD,beside=T,names.arg=indlist,col=c(nicecol),las=2)\n",
    "\n",
    "barplot(numsA,beside=T,names.arg=indlist,col=c(nicecol),las=2)\n",
    "barplot(numsD,beside=T,names.arg=indlist,col=c(nicecol),las=2)\n",
    "\n",
    "```\n",
    "\n",
    "Or both together:\n",
    "\n",
    "```\n",
    "sepcol=c(\"blue\",\"orange\")\n",
    "barplot(rbind(lengsA,lengsD),beside=T,names.arg=rbind(indlist,indlist),col=sepcol,las=2)\n",
    "barplot(rbind(numsA,numsD),beside=T,names.arg=rbind(indlist,indlist),col=sepcol,las=2)\n",
    "```\n",
    "\n",
    "Solution for more pretty violinplots:\n",
    "\n",
    "```\n",
    "library(ggplot2)\n",
    "mtab<-data.frame(lengs=c(lengsA,lengsD),sample=c(rep(\"AltaiNea\",length(lengsA)),rep(\"Denisovan\",length(lengsD))))\n",
    "p1<-ggplot(mtab,aes(factor(sample), lengs)) + \n",
    "    geom_violin(scale=\"width\",aes(fill=factor(sample)),adjust=1.0,draw_quantiles = c(0.5)  ) +\n",
    "    geom_jitter(height = 0, width = 0.25,size=1) + geom_segment(aes(x=0.0,y=0,xend=0.0,yend=1))    + \n",
    "    theme_bw() +\n",
    "    theme(panel.border = element_blank(),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),plot.background = element_rect(fill = \"white\"),legend.position=\"none\",axis.text.y= element_text(size=8),axis.title.y= element_text(size=10), axis.ticks.x = element_blank(), plot.title = element_text(face=\"bold\",hjust=0.5,size=12)) +\n",
    "    xlab(\"\")+ylab(\"Length distributions\") + ggtitle(label=\"Distribution of cumulative fragment lengths\")\n",
    "\n",
    "mtab2<-data.frame(lengs=c(numsA,numsD),sample=c(rep(\"AltaiNea\",length(lengsA)),rep(\"Denisovan\",length(lengsD))))\n",
    "p2<-ggplot(mtab2,aes(factor(sample), lengs)) + \n",
    "    geom_violin(scale=\"width\",aes(fill=factor(sample)),adjust=1.0,draw_quantiles = c(0.5)  ) +\n",
    "    geom_jitter(height = 0, width = 0.25,size=1) + geom_segment(aes(x=0.0,y=0,xend=0.0,yend=1))    + \n",
    "    theme_bw() +\n",
    "    theme(panel.border = element_blank(),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),plot.background = element_rect(fill = \"white\"),legend.position=\"none\",axis.text.y= element_text(size=8),axis.title.y= element_text(size=10), axis.ticks.x = element_blank(), plot.title = element_text(face=\"bold\",hjust=0.5,size=12)) +\n",
    "    xlab(\"\")+ylab(\"Length distributions\") + ggtitle(label=\"Distribution of fragment numbers\")\n",
    "\n",
    "install.packages(\"patchwork\")\n",
    "require(\"patchwork\")\n",
    "p1 + p2\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25aac508",
   "metadata": {},
   "source": [
    "## D) \n",
    "\n",
    "Creating a countour plot, basically taken from here: https://github.com/YingZhou001/sprimepipeline/blob/main/pub.pipeline.pbs/tools/plot_contour.r\n",
    "\n",
    "```\n",
    "library(MASS)\n",
    "\n",
    "matchtab1<-read.table(\"altai.final.bed\",sep=\"\\t\",header=F)\n",
    "matchtab2<-read.table(\"deni.final.bed\",sep=\"\\t\",header=F)\n",
    "matab<-data.frame(AltaiNea=matchtab1[,5],Deni=matchtab2[,5])\n",
    "\n",
    "png(\"contour.png\",width=1100,height=1100, res=250); \n",
    "ylab=\"Match to Denisovan\"; \n",
    "xlab=\"Match to Altai Neanderthal\"; \n",
    "level1=seq(0.3,0.9,0.1); \n",
    "level2=seq(1,10,1)\n",
    "X<-matab$AltaiNea\n",
    "Y<-matab$Deni\n",
    "o<-!is.na(X)&!is.na(Y)\n",
    "X<-X[o]\n",
    "Y<-Y[o]\n",
    "\n",
    "contour(kde2d(X,Y,n=100, lims=c(0,1,0,1)),levels=c(level1,level2), xaxs=\"i\",yaxs=\"i\",xlab=xlab,ylab=ylab,main=\"\",las=1,cex.lab=1.3,cex.axis=1.2,cex.main=1.5,lty=5,labcex=0.8);grid(lty=3,col=\"gray\")\n",
    "dev.off()\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a2bdbf2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
