{"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.7.12","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":5,"nbformat":4,"cells":[{"cell_type":"markdown","source":"## Introduction\n\n### What is introgression\n\nIntrogression is the phenomenon that genetic materials move from one lineage into another lineage between two different species.\n\n<img src=\"https://upload.wikimedia.org/wikipedia/commons/thumb/b/b8/Homo_sapiens_lineage.svg/1200px-Homo_sapiens_lineage.svg.png\" alt=\"human_introgression\" width=\"500\"/>\n\n**Figure 1 Interbreeding between archaic and modern humans.** This figure is from https://en.wikipedia.org/wiki/Interbreeding_between_archaic_and_modern_humans.\n\n### Strategies for indentifying introgressed regions\n\nThere are two major strategies to indentify introgressed regions.\n1. If we have genomes from source populations, for example, genomes from Neanderthals, we can compare modern human genomes with Neanderthal genomes directly.\n2. If we do not have source genomes, for example, we want to indentify introgression from an unknown lineage (ghost introgression), we can apply some statistical models to find out unusal regions.\n\nSeveral tools can indentify introgression without source populations.\n- [SkovHMM](https://github.com/LauritsSkov/Introgression-detection)\n- [sstar](https://github.com/admixVIE/sstar)\n- [SPrime](https://github.com/browning-lab/sprime)\n\nIn this tutorial, we will try SPrime.\n\n## SPrime pipeline\n\n<img src=\"https://ars.els-cdn.com/content/image/1-s2.0-S2666166721002574-fx1.jpg\" alt=\"sprimepipeline\"/>\n\n**Figure 2 The SPrime pipeline.** This figure is from [Zhou and Browing (2021)](https://doi.org/10.1016/j.xpro.2021.100550).\n\n\n### Download SPrime\n\nWe can download `SPrime` using the command `wget`.\n\n```\nwget https://faculty.washington.edu/browning/sprime.jar\n```\n\n### Run SPrime\n\nHere, we provide two datasets for detecting introgressed fragments in the CEU population with `SPrime`.\n1. The chromosome 21 with biallelic SNPs from the YRI and CEU populations.\n2. The TLR cluster in the chromosome 4 with biallelic SNPs from the YRI and CEU populations.\n\nThen we can use the following commands.\n\n```\njava -Xmx1g -jar sprime.jar gt=./data/chr21.YRI.CEU.biallelic.snps.vcf.gz outgroup=./data/YRI.list map=./data/plink.chr21.GRCh37.map out=chr21.CEU.introgressed\n```\nand\n```\njava -Xmx1g -jar sprime.jar gt=./data/chr4.YRI.CEU.TLR.biallelic.snps.vcf.gz outgroup=./data/YRI.list map=./data/plink.chr4.GRCh37.map out=chr4.CEU.TLR.introgressed\n```\n\n- The argument `gt` specifies a VCF file containing the genetic data we want to use.\n- The argument `outgroup` specifies a file containing samples without introgression (we assume), for example, the YRI population.\n- The argument `map` specifies a file containing a genetic map associated with the VCF file we want to use.\n- The argument `out` specifies the prefix of the output files.\n\nOther arguments in `SPrime` can be found in its [manual](https://github.com/browning-lab/sprime) or use the command `java -jar sprime.jar`.\n\n### Process SPrime output\n\nThere are two output files for each dataset after running `SPrime`. One ended with `.log` contains a summary of the analysis from `SPrime`, the other one ended with `.score` contains the results we want to have a look at.\n\nAn example for the `score` file is below.\n\n```\nCHROM   POS     ID      REF     ALT     SEGMENT ALLELE  SCORE\n21      14410518        rs186656431     T       C       2       1       470440\n21      14415190        rs199840792     G       T       2       1       470440\n21      14426851        rs138088002     T       C       2       1       470440\n```\n- The `CHROM` column is the name of the chromosome.\n- The `POS` column is the position of the variant.\n- The `ID` column is the name of the variant.\n- The `REF` column is the reference allele of the variant.\n- The `ALT` column is the alternative allele of the variant.\n- The `SEGMENT` column is the index of the introgressed segment.\n- The `ALLELE` column is the introgressed allele.\n- The `SCORE` column is the *S'* score.\n\nHowever, we want to convert the score file into [BED format](https://genome.ucsc.edu/FAQ/FAQformat.html#format1), which is more easier to be visualized.\n\nHere we provide a customized `Python` script to convert a score file into a BED file with the following commands.\n\n```\npython process_sprime_output.py chr21.CEU.introgressed.score chr21.CEU.introgressed.bed\n```\nand\n```\npython process_sprime_output.py chr4.CEU.TLR.introgressed.score chr4.CEU.TLR.introgressed.bed\n```\n\n### Check results\n\nWe can simply print out `chr4.CEU.TLR.introgressed.bed` in the terminal with the command `cat chr4.CEU.TLR.introgressed.bed`.\n\n```\n4       38760338        38905731        0       239319\n```\n- The first column is the name of the chromosome.\n- The second column is the start position of the introgressed region.\n- The third column is the ending position of the introgressed region.\n- The fourth column is the index of the introgressed segment from `SPrime`.\n- the fifth column is the *S'* score estimated by `SPrime`. If this score is larger than a threshold (the default is 100,000), then `SPrime` would assign the region as a introgressed region.\n\nThis is the TLR cluster under adaptive introgression suggested by [Dannemann et al. (2015)](http://dx.doi.org/10.1016/j.ajhg.2015.11.015).\n\n<img src=\"https://els-jbs-prod-cdn.jbs.elsevierhealth.com/cms/attachment/6110f2d8-da32-4138-85bb-13e4e79c7b05/gr1.jpg\" />\n\n**Figure 3 The introgressed region encompassing the genes *TLR10*, *TLR1*, and *TLR6* on chromosome 4.** This figure is from [Dannemann et al. (2015)](http://dx.doi.org/10.1016/j.ajhg.2015.11.015).\n\nFor the chromosome 21, we can use [the UCSC Genome Browser](https://genome.ucsc.edu/) to visualize the results.","metadata":{},"id":"16824698"}]}