{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "30cbefa4",
   "metadata": {},
   "source": [
    "# eigenstrat format dataset\n",
    "\n",
    "First, we need to have our genetic data! We will look at a dataset with 104 populations (total of 198 individuals), with data for 1.24 million positions across the human genome.\n",
    "\n",
    "We can have a look at the top few lines of each file to understand what is inside:\n",
    "\n",
    "    head dataset.ind\n",
    "\n",
    "    head dataset.snp\n",
    "\n",
    "We can also check the length of each of those files, which tell us how many individuals and how many SNPs are in the dataset:\n",
    "\n",
    "```\n",
    "wc -l dataset.ind\n",
    "wc -l dataset.snp\n",
    "```\n",
    "\n",
    "The \"dataset.geno\" file is in binary format and therefore can't be visualised as text, but it contains the genotype information for each individual, for each 1.24 million positions.\n",
    "\n",
    "We included a total of 74 modern populations from all across the world, and the following 28 ancient individuals:\n",
    "\n",
    "```\n",
    "    Altai_Neanderthal.DG\n",
    "    Austria_Krems1_1\n",
    "    Belgium_UP_GoyetQ116_1\n",
    "    Belgium_UP_GoyetQ_2\n",
    "    Chagyrskaya_Neandert.SG\n",
    "    China_Tianyuan\n",
    "    Czech_Vestonice16\n",
    "    Denisova.DG\n",
    "    DenisovaNeanderthalMix.SG\n",
    "    Goyet_Neanderthal.SG\n",
    "    Greenland_Saqqaq.SG\n",
    "    LesCottes_Neanderthal.SG\n",
    "    Luxembourg_Loschbour.DG\n",
    "    Mezmaiskaya1_Neander.SG\n",
    "    Mezmaiskaya2_Neander.SG\n",
    "    Romania_Muierii\n",
    "    Romania_Oase\n",
    "    Russia_HG_Karelia.SG\n",
    "    Russia_Kostenki14.SG\n",
    "    Russia_MA1_HG.SG\n",
    "    Russia_Sunghir3.SG\n",
    "    Russia_Ust_Ishim.DG\n",
    "    Spain_ElMiron\n",
    "    Spy_Neanderthal.SG\n",
    "    Turkey_Epipaleolithic\n",
    "    USA_Anzick.SG\n",
    "    USA_WA_Kennewick.SG\n",
    "    Vindija.DG\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "163c7003",
   "metadata": {},
   "source": [
    "## f4-statistics \n",
    "\n",
    "f4-statistics can be used to investigate if modern populations (X) shared more alleles with Neanderthals versus Denisovans (or vice versa), a classical marker of admixture.\n",
    "\n",
    "```\n",
    "f4(Yoruba, X; Neanderthal, Denisovan)\n",
    "f4 > 0      X closer to Denisovan than to Neanderthal\n",
    "f4 < 0      X closer to Neanderthal than to Denisovan\n",
    "```\n",
    "\n",
    "To actually calculate these statistics, the software \"qpDstat\" from the ADMIXTOOLS package is used.\n",
    "\n",
    "To make life easier, we will use admixtools in R, which provides the same and more functionalities.\n",
    "\n",
    "```\n",
    "R --vanilla\n",
    "library(admixtools)\n",
    "```\n",
    "\n",
    "\n",
    "wget https://ucloud.univie.ac.at/index.php/s/fYNi6xXR83R1YEc/download\n",
    "\n",
    "tar -zxvf download\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "This program requires a parameter file with all the information in it. This includes the name of our files and dataset, as well as a list of f4 tests we want to run. Let's open one of the provided parameter files.\n",
    "\n",
    "    nano f4_parameters_QinStoneking_1a\n",
    "\n",
    "This is what it looks like, and what each section represents:\n",
    "\n",
    "    genotypename: iNEAL_f4v2_PAM.geno\n",
    "    snpname: iNEAL_f4v2_PAM.snp\n",
    "    indivname: iNEAL_f4v2_PAM.ind\n",
    "    popfilename: f4_tests_QinStoneking_1a\n",
    "    f4mode: YES\n",
    "    details: YES\n",
    "\n",
    "Note how a specific popfilename is already mentioned there (\"f4_tests_QinStoneking_1a\"). That is the first file with the f4-testes we want to run, and it looks like this:\n",
    "\n",
    "    Yoruba.DG Spanish.DG Altai_Neanderthal.DG Denisova.DG\n",
    "    Yoruba.DG French.DG Altai_Neanderthal.DG Denisova.DG\n",
    "    Yoruba.DG Tuscan_1.DG Altai_Neanderthal.DG Denisova.DG\n",
    "    Yoruba.DG Finnish.DG Altai_Neanderthal.DG Denisova.DG\n",
    "\n",
    "Here, our X are the Spanish, French, Tuscan, and Finish populations. There are two more similar files to this one, where the tests were separated into small groups so that the Binder system doesn't run out of memory.\n",
    "\n",
    "We now have everything we need to run our first f4-statistics!\n",
    "\n",
    "    qpDstat -p f4_parameters_QinStoneking_1a\n",
    "    \n",
    "The output information we want to look at is the following:\n",
    "\n",
    "                PopA      PopB           PopC             PopD            f4            Z     BABA    ABBA   SNPS\n",
    "    result:  Yoruba.DG Spanish.DG Altai_Neanderthal.DG Denisova.DG     -0.001129     -5.533   21757  22938 1045961 \n",
    "    result:  Yoruba.DG  French.DG Altai_Neanderthal.DG Denisova.DG     -0.001194     -6.099   21793  23042 1046082 \n",
    "    result:  Yoruba.DG Tuscan_1.DG Altai_Neanderthal.DG Denisova.DG     -0.001421     -7.053   21697  23184 1046007 \n",
    "    result:  Yoruba.DG Finnish.DG Altai_Neanderthal.DG Denisova.DG     -0.001397     -7.087   21687  23148 1046076 \n",
    "\n",
    "And here is what each column represents:\n",
    "\n",
    "    f4 - The f4-statistic value\n",
    "    Z - The Z-score, which provides statistical validation of the significance of the f4\n",
    "    BABA - BABA counts for similarities between A/C or B/D\n",
    "    ABBA - ABBA counts for similarities between A/D or B/C\n",
    "    SNPS - Number of SNPs used for this statistic (overlapping SNPs between all 4 populations)\n",
    "\n",
    "Now let's run the other two commands and get the remaining results.\n",
    "\n",
    "    qpDstat -p f4_parameters_QinStoneking_1b\n",
    "    qpDstat -p f4_parameters_QinStoneking_1c\n",
    "\n",
    "And here are all the results combined:\n",
    "\n",
    "                PopA      PopB           PopC             PopD            f4            Z     BABA    ABBA   SNPS\n",
    "    result:  Yoruba.DG Spanish.DG Altai_Neanderthal.DG Denisova.DG     -0.001129     -5.533   21757  22938 1045961 \n",
    "    result:  Yoruba.DG  French.DG Altai_Neanderthal.DG Denisova.DG     -0.001194     -6.099   21793  23042 1046082 \n",
    "    result:  Yoruba.DG Tuscan_1.DG Altai_Neanderthal.DG Denisova.DG     -0.001421     -7.053   21697  23184 1046007 \n",
    "    result:  Yoruba.DG Finnish.DG Altai_Neanderthal.DG Denisova.DG     -0.001397     -7.087   21687  23148 1046076 \n",
    "\n",
    "    result:  Yoruba.DG Armenian.DG Altai_Neanderthal.DG Denisova.DG     -0.001363     -6.694   21642  23067 1046017 \n",
    "    result:  Yoruba.DG Iranian.DG Altai_Neanderthal.DG Denisova.DG     -0.001199     -6.250   21631  22886 1046045 \n",
    "    \n",
    "    result:  Yoruba.DG Quechua.DG Altai_Neanderthal.DG Denisova.DG     -0.001498     -6.652   21665  23233 1046011 \n",
    "    result:  Yoruba.DG   Mayan.DG Altai_Neanderthal.DG Denisova.DG     -0.001312     -5.913   21833  23205 1045905 \n",
    "    result:  Yoruba.DG Karitiana.DG Altai_Neanderthal.DG Denisova.DG     -0.001316     -5.770   21834  23210 1046048 \n",
    "\n",
    "    result:  Yoruba.DG Tibetan.DG Altai_Neanderthal.DG Denisova.DG     -0.001227     -5.802   21743  23027 1046002 \n",
    "    result:  Yoruba.DG Sindhi_Pakistan.DG Altai_Neanderthal.DG Denisova.DG     -0.001005     -5.259   21684  22735 1045979 \n",
    "    result:  Yoruba.DG     Han.DG Altai_Neanderthal.DG Denisova.DG     -0.001232     -5.907   21781  23070 1046067 \n",
    "    result:  Yoruba.DG     Dai.DG Altai_Neanderthal.DG Denisova.DG     -0.001300     -6.733   21652  23012 1046097 \n",
    "    result:  Yoruba.DG   Yakut.DG Altai_Neanderthal.DG Denisova.DG     -0.001538     -7.240   21553  23162 1045991 \n",
    "    result:  Yoruba.DG  Papuan.DG Altai_Neanderthal.DG Denisova.DG     -0.000076     -0.297   22618  22698 1046060 \n",
    "    result:  Yoruba.DG    Onge.DG Altai_Neanderthal.DG Denisova.DG     -0.001469     -6.741   21427  22963 1045475 \n",
    "\n",
    "## Replicating Hajdinjak et al. 2021 ##\n",
    "\n",
    "Hajdinjak and colleagues used f4-statistics to investigate if some ancient/archaic individuals (X) shared more alleles with the eastern Tianyuan or western Kostenki14.\n",
    "With this they wanted to see if those individuals contributed more towards modern East or West Eurasians.\n",
    "\n",
    "    D/f4(Tianyuan, Kostenki14; X, Mbuti) \n",
    "    f4(A, B; C, D) or f4(D, C; B, A)\n",
    "\n",
    "    f4(Tianyuan, Kostenki14; X, Mbuti) \n",
    "    f4(Mbuti, X; Kostenki14, Tianyuan)\n",
    "\n",
    "    D/f4 > 0        X closer to Tianyuan than to Kostenki14\n",
    "    D/f4 < 0        X closer to Kostenki14 than to Tianyuan\n",
    "\n",
    "We separated our data again, this time into 2 sets of files, so let's run the first one, then the second:\n",
    "\n",
    "    qpDstat -p f4_parameters_Hajdinjak_2a\n",
    "    qpDstat -p f4_parameters_Hajdinjak_2b\n",
    "\n",
    "And here are the results:\n",
    "\n",
    "    result: China_Tianyuan Russia_Kostenki14.SG Russia_Sunghir3.SG   Mbuti.DG     -0.012435    -15.004   41016  51074 808838 \n",
    "    result: China_Tianyuan Russia_Kostenki14.SG Czech_Vestonice16   Mbuti.DG     -0.010691    -13.274   31323  37910 616144 \n",
    "    result: China_Tianyuan Russia_Kostenki14.SG Belgium_UP_GoyetQ116_1   Mbuti.DG     -0.006486     -7.652   35397  39666 658252 \n",
    "    result: China_Tianyuan Russia_Kostenki14.SG Luxembourg_Loschbour.DG   Mbuti.DG     -0.008428    -11.930   42232  48976 800247 \n",
    "    result: China_Tianyuan Russia_Kostenki14.SG Russia_HG_Karelia.SG   Mbuti.DG     -0.004793     -6.619   38305  41683 704840 \n",
    "    \n",
    "    result: China_Tianyuan Russia_Kostenki14.SG Romania_Oase   Mbuti.DG      0.001606      1.760    9108   8835 169768 \n",
    "    result: China_Tianyuan Russia_Kostenki14.SG Russia_Ust_Ishim.DG   Mbuti.DG     -0.000333     -0.449   42335  42590 766348 \n",
    "    result: China_Tianyuan Russia_Kostenki14.SG USA_Anzick.SG   Mbuti.DG      0.003994      5.845   47439  44215 807327 \n",
    "    result: China_Tianyuan Russia_Kostenki14.SG USA_WA_Kennewick.SG   Mbuti.DG      0.003807      5.091   24728  23106 425924 \n",
    "    result: China_Tianyuan Russia_Kostenki14.SG Greenland_Saqqaq.SG   Mbuti.DG      0.005727      7.831   46410  41922 783674\n",
    "\n",
    "## Replicating Mafessoni et al. 2020 ##\n",
    "\n",
    "Mafessoni and colleagues tried to understand two things about published archaic individuals and their relationship towards Chagyrskaya 8:\n",
    "\n",
    "a) were they closer to the newly presented 80 kyo Chagyrskaya individual, than to the 100 kyo Neanderthal from the nearby 100km Denisova Cave\n",
    "b) or instead were they closer to the 50 kyo Vindija individual.\n",
    "\n",
    "Again, we divided the data into 2 parts. Part 3a contains the tests against Altai_Neanderthal (AKA Denisova 5), and part 3b against Vindija33.19.\n",
    "\n",
    "    qpDstat -p f4_parameters_Mafessoni_3a\n",
    "    qpDstat -p f4_parameters_Mafessoni_3b\n",
    "\n",
    "Here are the results for 3a (100 kyo Denisova 5 vs 80 kyo Chagyrskaya 8):\n",
    "\n",
    "    result:  Chimp.REF LesCottes_Neanderthal.SG Altai_Neanderthal.DG Chagyrskaya_Neandert.SG      0.003413     18.195    6130   3330 820204 \n",
    "    result:  Chimp.REF Spy_Neanderthal.SG Altai_Neanderthal.DG Chagyrskaya_Neandert.SG      0.003340     16.411    3538   1917 485146 \n",
    "    result:  Chimp.REF Goyet_Neanderthal.SG Altai_Neanderthal.DG Chagyrskaya_Neandert.SG      0.003582     17.743    6059   3218 793119 \n",
    "    result:  Chimp.REF Mezmaiskaya1_Neander.SG Altai_Neanderthal.DG Chagyrskaya_Neandert.SG      0.003322     16.482    4375   2400 594502 \n",
    "    result:  Chimp.REF Vindija.DG Altai_Neanderthal.DG Chagyrskaya_Neandert.SG      0.003569     18.970    7759   4117 1020387 \n",
    "    result:  Chimp.REF DenisovaNeanderthalMix.SG Altai_Neanderthal.DG Chagyrskaya_Neandert.SG      0.001598     11.833    4957   3565 871328 \n",
    "    result:  Chimp.REF Denisova.DG Altai_Neanderthal.DG Chagyrskaya_Neandert.SG     -0.000282     -2.153    3941   4229 1020023 \n",
    "\n",
    "And here are the results for 3b (50 kyo Vindija33.19 vs 80 kyo Chagyrskaya 8):\n",
    "\n",
    "    result:  Chimp.REF LesCottes_Neanderthal.SG Vindija.DG Chagyrskaya_Neandert.SG     -0.001415     -9.747    2940   4100 820223 \n",
    "    result:  Chimp.REF Spy_Neanderthal.SG Vindija.DG Chagyrskaya_Neandert.SG     -0.001927    -11.803    1582   2518 485176 \n",
    "    result:  Chimp.REF Goyet_Neanderthal.SG Vindija.DG Chagyrskaya_Neandert.SG     -0.002280    -14.176    2584   4393 793141 \n",
    "    result:  Chimp.REF Mezmaiskaya1_Neander.SG Vindija.DG Chagyrskaya_Neandert.SG     -0.000041     -0.292    2522   2546 594520 \n",
    "    result:  Chimp.REF Altai_Neanderthal.DG Vindija.DG Chagyrskaya_Neandert.SG      0.000039      0.306    4158   4117 1020387 \n",
    "    result:  Chimp.REF DenisovaNeanderthalMix.SG Vindija.DG Chagyrskaya_Neandert.SG      0.000197      1.800    3242   3070 871324 \n",
    "    result:  Chimp.REF Denisova.DG Vindija.DG Chagyrskaya_Neandert.SG     -0.000022     -0.228    3001   3024 1020022 \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69f7d977",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
