{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "24435468",
   "metadata": {},
   "source": [
    "# 3 Filetypes\n",
    "\n",
    "* How to count the number of reads?\n",
    "\n",
    "```\n",
    "grep \"^@\" sample_seq.fq\n",
    "```\n",
    "\n",
    "\n",
    "* How to do adapter trimming?\n",
    "\n",
    "```\n",
    "module load trimmomatic\n",
    "\n",
    "trimmomatic SE -phred33 sample_seq.fq sample_seq_p.fq \\\n",
    "LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
    "```\n",
    "\n",
    "\n",
    "* How to filter mappped sequencing reads?\n",
    "\n",
    "```\n",
    "samtools view -b -q 30 -F 4 -o sample.sorted.filtered.bam sample.sorted.bam\n",
    "samtools index sample.sorted.filtered.bam\n",
    "\n",
    "gatk MarkDuplicates --REMOVE_DUPLICATES -I sample.sorted.filtered.bam -O sample.sorted.filtered.rmdup.bam -M metrics.txt\n",
    "samtools index sample.sorted.filtered.rmdup.bam\n",
    "```\n",
    "\n",
    "\n",
    "* GATK\n",
    "\n",
    "Problem encountered = `bam` file is not indexed. Solved by: `samtools index sample.sorted.bam`.\n",
    "\n",
    "\n",
    "* How many SNPs are the in the file?\n",
    "\n",
    "This is a zipped file. Use `zcat sample.vcf.gz | ` to run standard commands on it.\n",
    "\n",
    "```\n",
    "zcat sample.vcf.gz | grep -v \"^#\" | wc -l\n",
    "```\n",
    "\n",
    "\n",
    "* What is the position of the first and the last SNP?\n",
    "\n",
    "```\n",
    "zcat sample.vcf.gz | tail -n 1 | cut -f2\n",
    "```\n",
    "\n",
    "\n",
    "* Here is another vcf file. Can you see some differences to the one you just created?\n",
    "\n",
    "The main difference is that one contains information for each position on the chromosome, the other only for positions that are not idential to the reference genomes (variable positions).\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13397411",
   "metadata": {},
   "source": [
    "# 4 Data filtering\n",
    "\n",
    "* Copy them to your directory and index them!\n",
    "\n",
    "```\n",
    "cp $main/share/altai.filt.vcf.gz $main/martin/\n",
    "cp $main/share/denisovan.filt.vcf.gz $main/martin/\n",
    "\n",
    "bcftools index denisovan.filt.vcf.gz\n",
    "bcftools index altai.filt.vcf.gz\n",
    "```\n",
    "\n",
    "\n",
    "* Merge them together with the `chr21.YRI.CEU.vcf.gz`!\n",
    "\n",
    "```\n",
    "bcftools merge -Oz -o merged_file.vcf.gz -g $refgenome chr21.YRI.CEU.vcf.gz altai.filt.vcf.gz denisovan.filt.vcf.gz\n",
    "```\n",
    "\n",
    "\n",
    "* Extra task: Remove sites that have missing data in any individual, and count the number of sites for which this is true.\n",
    "\n",
    "```\n",
    "bcftools filter '-e GT==\"./.\"' -o merged_file_nomiss.vcf.gz merged_file.vcf.gz \n",
    "bcftools filter '-i GT==\"./.\"'  merged_file.vcf.gz | bcftools view -H | wc -l \n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0200ded4",
   "metadata": {},
   "source": [
    "# 5 Data analysis\n",
    "\n",
    "\n",
    "* Read in the summary table\n",
    "```\n",
    "hets<-read.table(\"hets.txt\", sep=\"\\t\",header=T,comment.char=\"\")\n",
    "```\n",
    "\n",
    "\n",
    "* Plot a histogram\n",
    "\n",
    "```\n",
    "png(\"hets_histogram.png\",600,600)\n",
    "hist(hets[,6],breaks=20)\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "\n",
    "* Calculate the mean for CEU and YRI separately, compared to the archaics\n",
    "\n",
    "```\n",
    "ceu<-unlist(read.table(\"appladmix/notebooks/1_data/data/CEU.list\", sep=\"\\t\",header=F))\n",
    "yri<-unlist(read.table(\"appladmix/notebooks/1_data/data/YRI.list\", sep=\"\\t\",header=F))\n",
    "mean(hets[which(hets[,3]%in%ceu),6])\n",
    "mean(hets[which(hets[,3]%in%yri),6])\n",
    "mean(hets[which(hets[,3]%in%c(\"AltaiNea\",\"DenisovaPinky\")),6])\n",
    "```\n",
    "\n",
    "\n",
    "* Perform a statistical test whether CEU and YRI are different\n",
    "\n",
    "```\n",
    "wilcox.test(hets[which(hets[,3]%in%ceu),6],hets[which(hets[,3]%in%yri),6])\n",
    "```\n",
    "\n",
    "\n",
    "* There are ~30 million bases for which we can have data on chr21. Often heterozygosity is expressed als hets per 1,000bp. Calculate this and  present the values in such a way.\n",
    "\n",
    "```\n",
    "genom=30000000\n",
    "\n",
    "mean(hets[which(hets[,3]%in%ceu),6])/genom*1000\n",
    "mean(hets[which(hets[,3]%in%yri),6])/genom*1000\n",
    "hets[which(hets[,3]%in%c(\"AltaiNea\",\"DenisovaPinky\")),6]/genom*1000\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6890ed7a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
