{"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.7.12","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":5,"nbformat":4,"cells":[{"cell_type":"markdown","source":"## Introduction\n\n### What is natural selection\n\nNatural selection is the phenomenon that individuals in a population survive and reproduce differently because of differences in their pheotypes.\n\nThere are different types of natural selection, for example,\n- Purifying selection\n- Positive selection\n\n<img src=\"https://d3i71xaburhd42.cloudfront.net/326581abd3371faf63fad44f6c31b833854cb37f/3-Figure1-1.png\" width=\"800\"/>\n\n**Figure 1 Genomic signatures of positive selection.** This figure is from [Biswas and Akey (2006)](https://doi.org/10.1016/j.tig.2006.06.005).\n\n### Extended haplotype homozygosity (EHH) and integrated haplotype score (iHS)\n\nEHH is a statistic to measure the decrease of variation around a SNP. This is because positive selection could decrease the diversity around a variant under selection, as shown in Figure 1.\n\niHS is a measure of the amount of (EHH) at a given SNP along the ancestral allele relative to the derived allele.\n\n<img src=\"https://journals.plos.org/plosbiology/article/figure/image?size=medium&id=10.1371/journal.pbio.0040072.g006\">\n\n**Figure 2 Signals of selection for three candidate selection regions using EHH and iHS.** This figure is from [Voight et al. (2006)](https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.0040072).\n\n\n## selscan pipeline\n\nIn this tutorial, we will use `selscan` to calculate the iHS scores in the *LCT* region from the CEU population.\n\nThe *LCT* gene makes an enzyme called lactase to digest lactose. It has the strongest signals of positive selection in Europeans ([Gerbault 2014](https://doi.org/10.1159/000360136)).\n\n<img src=\"https://upload.wikimedia.org/wikipedia/commons/thumb/f/ff/Lactose_tolerance_in_the_Old_World.svg/1280px-Lactose_tolerance_in_the_Old_World.svg.png\" width=\"500\">\n\n**Figure 3 An estimate of the percentage of adults that can digest lactose in the indigenous population of the Old World.** This figure is from https://en.wikipedia.org/wiki/Lactose_intolerance.\n\n### Download selscan\n\nWe can download `selscan` with `git clone`.\n\n```\ngit clone https://github.com/szpiech/selscan\n```\n\n### Run selscan\n\nHere, we provide a VCF file containig the *LCT* gene with 1 Mb flank size from the CEU population and a map file containing the information of each variant in the VCF file.\n\nThen we can calculate the iHS scores in this dataset with the following command.\n\n```\n./selscan/bin/linux/selscan --vcf ./data/chr2.CEU.LCT.flank.1mb.biallelic.snps.vcf.gz --map ./data/chr2.CEU.LCT.flank.1mb.biallelic.snps.map --ihs\n```\n\nIf you are using macOS, then use\n\n```\n./selscan/bin/osx/selscan --vcf ./data/chr2.CEU.LCT.flank.1mb.biallelic.snps.vcf.gz --map ./data/chr2.CEU.LCT.flank.1mb.biallelic.snps.map --ihs\n```\n\n### Check results\n\nThere are two output files from `selscan`. One is the log file storing the summary of the analysis. We should look at the file ended with `.out`.\n\nAn example is in below.\n\n```\nrs16831345      135871914       0.0858586       206664  253975  -0.0895271\nrs2034840       135876308       0.0858586       206664  254076  -0.0897007\nrs55807439      135886744       0.0808081       232011  251582  -0.0351707\nrs2305595       135887893       0.0858586       208931  254358  -0.0854418\n```\n- The first column is the name of the variant.\n- The second column is the position of the variant.\n- The third column is the frequency of the allele 1.\n- The fourth column is the iHH1 score.\n- The fifth column is the iHH0 score.\n- The sixth column is the unstandarized iHS score.\n\nMore details can be found in the [manual](https://github.com/szpiech/selscan/blob/master/manual/selscan-manual.pdf).\n\nTo explore the results quickly, we can simply sort the variants by their unstandarized iHS scores and choose those variants with top iHS scores as the candidate variants under selection.\n\n```\nsort -rnk 6,6 outfile.ihs.out | head -100\nsort -rnk 6,6 outfile.ihs.out | head -100 | awk '$1~/rs4988235/'\nsort -rnk 6,6 outfile.ihs.out | head -100 | awk '$1~/rs182549/'\n```\n\nFor example, we can find two classic SNPs [rs4988235](https://www.snpedia.com/index.php/Rs4988235) and [rs182549](https://www.snpedia.com/index.php/Rs182549) associated with lactose intolerance in Europeans are in the top candidates.","metadata":{},"id":"0d7cee04"}]}