{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "85c2d356-a24c-48c1-b1de-1e1c68e447dd",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Calculating a PCA in R\n",
    "\n",
    "Last time, we have calculated a PCA using a command line tool. There are many ways of doing it, and now we will do it in R as well.\n",
    "\n",
    "But let's first have a look at the output we generated last time and get a nicer output figure. Then, we will calculate it with R functions.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1e8e9b3-94df-4f4d-9b5a-d500161d03be",
   "metadata": {},
   "source": [
    "## Making a nicer plot\n",
    "\n",
    "I'm not happy with the plot, and maybe one could provide some info to the tool itself to get it in shape. But even better is to get a nice plot of your own. Let's `R` then!\n",
    "\n",
    "```\n",
    "R --vanilla\n",
    "```\n",
    "\n",
    "We read in the data and metadata:\n",
    "```\n",
    "pca<-read.table(\"geno_pca.eigenvec\",sep=\"\\t\", header=T,comment.char=\"\")\n",
    "ceu<-unlist(read.table(\"../CEU.list\", sep=\"\\t\",header=F,comment.char=\"\"))\n",
    "yri<-unlist(read.table(\"../YRI.list\", sep=\"\\t\",header=F,comment.char=\"\"))\n",
    "```\n",
    "\n",
    "We change groupings to our liking:\n",
    "```\n",
    "pca$Group[which(pca$SampleName%in%ceu)]<-\"European\"\n",
    "pca$Group[which(pca$SampleName%in%yri)]<-\"African\"\n",
    "pca$Group[which(pca$SampleName==\"AltaiNea\")]<-\"Neanderthal\"\n",
    "pca$Group[which(pca$SampleName==\"DenisovaPinky\")]<-\"Denisovan\"\n",
    "```\n",
    "\n",
    "And add a nice color coding:\n",
    "```\n",
    "pca$Cluster[which(pca$SampleName%in%ceu)]<-\"blue\"\n",
    "pca$Cluster[which(pca$SampleName%in%yri)]<-\"green\"\n",
    "pca$Cluster[which(pca$SampleName==\"AltaiNea\")]<-\"orange\"\n",
    "pca$Cluster[which(pca$SampleName==\"DenisovaPinky\")]<-\"red\"\n",
    "```\n",
    "\n",
    "And now make a nice plot:\n",
    "\n",
    "```\n",
    "pdf(\"PCA_chr21_col.pdf\",6,6)\n",
    "plot(pca[,4],pca[,5],col=pca[,3],pch=16,\n",
    "    xlab=\"PC1\",\n",
    "    ylab=\"PC2\",\n",
    "    main=paste(\"PCA of chr21\"))\n",
    "legend(\"bottomright\",legend=c(\"European\",\"African\",\"Neanderthal\",\"Denisovan\"), fill=c(\"blue\",\"green\",\"orange\",\"red\"),bty=\"n\" )\n",
    "dev.off()\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f51d1368",
   "metadata": {},
   "source": [
    "## A simple PCA in R\n",
    "\n",
    "### Data preparation\n",
    "\n",
    "Let's take the same file. What we need now is a genotype matrix, or a table of numbers which can be used by a statistical method. This table should also not have missing data, or other stuff such as non-biallelic SNVs. We may want to ensure this with another filtering step.\n",
    "\n",
    "Another neat feature of `bcftools` is that it actually can also transform `vcf` formatted files into anything you need. In this case, we really just want to have these numbers without anything else. `bcftools query` is good for that, since you can define which fields to extract, and how to separate them.\n",
    "\n",
    "* Let's do it in one go!\n",
    "\n",
    "```\n",
    "# in BASH\n",
    "\n",
    "bcftools view -a chr21.merged.vcf.gz | bcftools view -m 2 -M 2 -v snps | bcftools filter -e 'GT[*] = \"mis\"' | bcftools query -f '[%GT ]\\n' > allgts.txt\n",
    "```\n",
    "\n",
    "* This is still a very big file! Let's \"thin\" it a bit by randomly taking 100000 SNPs:\n",
    "\n",
    "```\n",
    "shuf -n 100000 allgts.txt > somegts.txt\n",
    "```\n",
    "\n",
    "There could be different ways of doing this, but here this way is convenient to keep it simple.\n",
    "\n",
    "* Then we can switch into `R`.\n",
    "\n",
    "\n",
    "### Data preparation in R\n",
    "\n",
    "```\n",
    "R --vanilla\n",
    "\n",
    "snps<-read.table(\"allgts.txt\", sep=\" \",header=F,comment.char=\"\")\n",
    "head(snps)\n",
    "```\n",
    "\n",
    "* Note that the separator is now a space, because that is how we defined it in the previous step. However, this was adding a last space to the end of the file, and R believes there is a (empty) column. Let's inspect the first and last column.\n",
    "\n",
    "```\n",
    "snps<-snps[,-ncol(snps)]\n",
    "\n",
    "table(snps[,1])\n",
    "table(snps[,209])\n",
    "```\n",
    "\n",
    "Now, these are not really numbers, but genotypes...\n",
    "\n",
    "* A very easy way to turn them into single numbers would be using the `ifelse` statement like this:\n",
    "\n",
    "```\n",
    "snps[,1]<-ifelse(snps[,1]%in%c(\"0|0\",\"0/0\"),0,snps[,1])\n",
    "snps[,1]<-ifelse(snps[,1]%in%c(\"1|1\",\"1/1\"),2,snps[,1])\n",
    "snps[,1]<-ifelse(snps[,1]%in%c(\"0|1\",\"1|0\",\"0/1\",\"1/0\"),1,snps[,1])\n",
    "table(snps[,1])\n",
    "```\n",
    "\n",
    "* However, this is not a numeric matrix:\n",
    "\n",
    "```\n",
    "is.numeric(snps[,1])\n",
    "snps[,1]<-as.numeric(snps[,1])\n",
    "is.numeric(snps[,1])\n",
    "```\n",
    "\n",
    "This looks better, but you certainly don't want to type in 208 more times the same things, and you would very likely make mistakes on the way.\n",
    "\n",
    "* Let's use a `for` loop and nested `ifelse` statements to make it easier! As a side-effect, the whole column will be numeric as well!\n",
    "\n",
    "```\n",
    "for (j in (2:ncol(snps))) { snps[,j]<-ifelse(snps[,j]%in%c(\"0|0\",\"0/0\"),0,ifelse(snps[,j]%in%c(\"1|1\",\"1/1\"),2,ifelse(snps[,j]%in%c(\"0|1\",\"1|0\",\"0/1\",\"1/0\"),1,NA))); print(j) }\n",
    "\n",
    "\n",
    "table(snps[,3])\n",
    "table(snps[,208])\n",
    "is.numeric(snps[,159])\n",
    "```\n",
    "\n",
    "I would note that at this stage it is ok to do things that way. Much better would be to write a `function` to make it more efficient...\n",
    "\n",
    "If you feel that coming up with efficient solutions is complicated, that is fine - it is a matter of practice and experience. We need to learn things in order to do them.\n",
    "\n",
    "\n",
    "### Calculating the PCA\n",
    "\n",
    "* Did we forget anything? Ah, yes, we may need the proper functions to calculate this! A possible (and very nice) libary with such functions [adegenet](https://adegenet.r-forge.r-project.org/).\n",
    "\n",
    "FYI: R has many packages, and often they are not pre-installed on a computer/structure. If that is the case, you may succeed with `install.packages(\"adegenet\")`. But fortunately, we have it available here, you don't need to!\n",
    "\n",
    "The method we want to use is called `dudi.pca`. It needs the individuals in the rows, and the positions in the columns, that is, a transposed table to be done with `t()`. As you see, nested functions are very common in `R`, do combine things on the fly without intermediate objects. \n",
    "\n",
    "* Let's to this:\n",
    "\n",
    "```\n",
    "library(\"adegenet\")\n",
    "\n",
    "pca_object <- dudi.pca(t(snps),nf=20,scannf=F)\n",
    "\n",
    "tb<-list()\n",
    "for (j in (1:ncol(snps)))  { tb[[j]]<-table(snps[,j]) }\n",
    "\n",
    "pca_object\n",
    "summary(pca_object)\n",
    "```\n",
    "\n",
    "This works so far! We have created a higher-level `R` object with lots of information. Important are the coordinates for each individual, which are in the sub-object called `li`, which we can access with the `$` sign.\n",
    "\n",
    "### Plotting the PCA\n",
    "\n",
    "* Let's plot this!\n",
    "\n",
    "```\n",
    "pdf(\"PCA_1000g.pdf\",12,12)\n",
    "plot(pca_object$li[,1],pca_object$li[,2])\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "Ok, this is just some distribution of dots... It is nicer to color it. For this, we need the metadata again. In `R`, we can then create a matching colour vector:\n",
    "\n",
    "```\n",
    "ceu<-unlist(read.table(\"../CEU.list\", sep=\"\\t\",header=F,comment.char=\"\"))\n",
    "yri<-unlist(read.table(\"../YRI.list\", sep=\"\\t\",header=F,comment.char=\"\"))\n",
    "names<-rbind(cbind(\"CEU\",ceu),cbind(\"YRI\",yri),cbind(c(\"AltaiNea\",\"DenisovaPinky\"),c(\"Neanderthal\",\"Denisovan))\n",
    "mypops<-c(\"CEU\",\"YRI\",\"Neanderthal\",\"Denisovan\")\n",
    "cols=c(\"blue\",\"green\",\"orange\",\"red\")\n",
    "\n",
    "#colvector<-names[,2]\n",
    "#for (pop in (1:nrow(mypops))) { colvector<-gsub(names[pop],cols[pop],colvector) }\n",
    "```\n",
    "\n",
    "* Let's plot it!\n",
    "\n",
    "```\n",
    "pdf(\"PCA_1000g_col.pdf\",12,12)\n",
    "plot(pca_object$li[,1],pca_object$li[,2],col=colvector)\n",
    "legend(\"topright\",legend=mypops, fill=cols,bty=\"n\" )\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "This is actually already quite good - considering that it is just a part of chromosome 18! We can do some more polishing to make it look nicer, and add the information how much of the variation is explained by these first two PCs.\n",
    "\n",
    "* Let's add nice axis labels (`xlab` and `ylab`), and points instead of empty circles (with `pch`):\n",
    "\n",
    "\n",
    "```\n",
    "percent_variation=round(pca_object$eig/sum(pca_object$eig)*100,2)\n",
    "pdf(\"PCA_1000g_nice.pdf\",12,12)\n",
    "plot(pca_object$li[,1],pca_object$li[,2],col=colvector,pch=16,\n",
    "    xlab=paste(\"PC1\",\" (\",percent_variation[1],\"%)\",sep=\"\"),\n",
    "    ylab=paste(\"PC\",2,\" (\",percent_variation[2],\"%)\",sep=\"\"),\n",
    "    main=paste(\"PCA of 1000G, chr18\"))\n",
    "legend(\"topright\",legend=mypops, fill=cols,bty=\"n\" )\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "Now, this looks good, and you can gain some insight into the relationships of these populations!\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45e09e3e",
   "metadata": {},
   "source": [
    "# Very good! Next step: D-statistics!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
