{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "466e22b6",
   "metadata": {},
   "source": [
    "# eigenstrat format dataset\n",
    "\n",
    "Now, everything is running. Let's first get our genetic data! We will look at a dataset with 104 populations (total of 198 individuals), with data for 1.24 million positions across the human genome.\n",
    "\n",
    "We get some data:\n",
    "\n",
    "```\n",
    "cd ~/notebooks/admixtools/data\n",
    "\n",
    "wget https://ucloud.univie.ac.at/index.php/s/fYNi6xXR83R1YEc/download\n",
    "tar -zxvf download\n",
    "```\n",
    "\n",
    "Let's have a look at the top few lines of each file to understand what is inside:\n",
    "\n",
    "    head dataset.ind\n",
    "\n",
    "    head dataset.snp\n",
    "\n",
    "Check the length of each of those files, which tell us how many individuals and how many SNPs are in the dataset:\n",
    "\n",
    "```\n",
    "wc -l dataset.ind\n",
    "wc -l dataset.snp\n",
    "```\n",
    "\n",
    "The \"dataset.geno\" file is in binary format and therefore can't be visualised as text, but it contains the genotype information for each individual, for each 1.24 million positions.\n",
    "\n",
    "A total of 74 modern populations from all across the world is included, and 28 ancient individuals, including some archaic hominins:\n",
    "\n",
    "```\n",
    "    Altai_Neanderthal.DG\n",
    "    Chagyrskaya_Neandert.SG\n",
    "    Denisova.DG\n",
    "    etc\n",
    "```\n",
    "\n",
    "These files are often interpreted with admixtools. To make life easier, we will use admixtools in R, which provides the same and more functionalities as the usual command line tools.\n",
    "\n",
    "You may use RStudio or call R from the command line, and load the package.\n",
    "\n",
    "```\n",
    "R --vanilla\n",
    "library(admixtools)\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd000b6c",
   "metadata": {},
   "source": [
    "## f4-statistics \n",
    "\n",
    "f4-statistics can be used to investigate allele sharing between populations, a classical marker of admixture. They are very similar to D-statistics.\n",
    "\n",
    "\n",
    "MK: picture!!\n",
    "\n",
    "For example if modern non-Africans (X) shared more alleles with Neanderthals than Africans:\n",
    "\n",
    "```\n",
    "f4(Yoruba, X; Neanderthal, Chimpanzee)\n",
    "f4 > 0      X closer to Denisovan than to Neanderthal\n",
    "f4 < 0      X closer to Neanderthal than to Denisovan\n",
    "```\n",
    "\n",
    "Or if a modern population (X) shared more alleles with Neanderthals than Denisovans.\n",
    "\n",
    "```\n",
    "f4(Yoruba, X; Neanderthal, Denisovan)\n",
    "f4 > 0      X closer to Denisovan than to Neanderthal\n",
    "f4 < 0      X closer to Neanderthal than to Denisovan\n",
    "```\n",
    "\n",
    "Now, the first step is to calculate the basic statistics. In order to do it within the Binder framework (large data), we need to restrict it to some individuals.\n",
    "\n",
    "There is a command to do this:\n",
    "\n",
    "```\n",
    "extract_f2(pref=\"dataset\",\n",
    "    outdir=\"genos\",blgsize=500000,\n",
    "    overwrite=T,maxmem=1000,\n",
    "    pops=c(\"Altai_Neanderthal.DG\",\"Denisova.DG\", \"Finnish.DG\", \"Japanese.DG\",\"Mbuti.DG\", \"Papuan.DG\",\"Yoruba.DG\",\"Chimp.REF\"))\n",
    "```\n",
    "\n",
    "This function will calculate pairwise statistics only for the samples in question, and write these into new files. Nothing much to see here, though you may inspect the new directory \"genos\" with some binary files.\n",
    "\n",
    "Now we load that into R:\n",
    "\n",
    "```\n",
    "f2_blocks = f2_from_precomp(\"genos\")\n",
    "```\n",
    "\n",
    "This is a complex table with the basic statistics we need. But we want to focus on f4-statistics to see if there was admixture between some populations. So, we calculate f4-statistics.\n",
    "\n",
    "First, let's check the f4-stats for Neanderthal introgression:\n",
    "\n",
    "```\n",
    "f4_table<-f4(f2_blocks,pop1=\"Yoruba.DG\", pop2=c(\"Finnish.DG\", \"Japanese.DG\",\"Mbuti.DG\", \"Papuan.DG\"),pop3=\"Altai_Neanderthal.DG\", pop4=\"Chimp.REF\")\n",
    "```\n",
    "\n",
    "What do we see here?\n",
    "\n",
    "```\n",
    "f4_table<-f4(f2_blocks,pop1=\"Yoruba.DG\", pop2=c(\"Finnish.DG\", \"Japanese.DG\",\"Mbuti.DG\", \"Papuan.DG\"),pop3=\"Altai_Neanderthal.DG\", pop4=\"Denisova.DG\")\n",
    "f4_table<-f4(f2_blocks,pop1=\"Japanese.DG\", pop2=c(\"Finnish.DG\", \"Japanese.DG\",\"Mbuti.DG\", \"Papuan.DG\"),pop3=\"Altai_Neanderthal.DG\", pop4=\"Denisova.DG\")\n",
    "```\n",
    "\n",
    "MK figure on history\n",
    "\n",
    "\n",
    "\n",
    "This is for all possible combinations!\n",
    "\n",
    "\n",
    "```\n",
    "f4_table[which(f4_table[,2]==\"Yoruba.DG\" & f4_table[,3]==\"Altai_Neanderthal.DG\" & f4_table[,4]==\"Chimp.REF\"),]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fd1d7bb-4556-410b-a279-5e6d8c36f6ff",
   "metadata": {},
   "source": [
    "## admixture graphs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
