{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c102ec68",
   "metadata": {},
   "source": [
    "# Population genetics\n",
    "\n",
    "Let's do some basic population genetics with the data we already have at hand here\n",
    "\n",
    "```\n",
    "module load bcftools R admixture htslib conda gatk\n",
    "conda activate plink2-2.0.0a.6.9 \n",
    "```\n",
    "\n",
    "We will use genomic data from all the 1000Genomes populations, in particular chromosome 18. We can define a variable with its location:\n",
    "\n",
    "```\n",
    "TGDATA=\"$main/share/ALL.chr18.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz\" \n",
    "\n",
    "ls $TGDATA\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66d9a421",
   "metadata": {},
   "source": [
    "## Heterozygosity\n",
    "\n",
    "Now, let's look at heterozygosity at a more global scale. Since the runtime and memory requirements of this algorithm , we select a subset of variants. There are several ways of doing this, here I demonstrate a simple way with `gatk`. Let's get 25% of the SNVs in the original VCF file:\n",
    "\n",
    "```\n",
    "gatk SelectVariants --select-random-fraction 0.25 --variant $TGDATA --output random_larger_subset.vcf.gz\n",
    "```\n",
    "\n",
    "Then, again with `bcftools stats`. \n",
    "\n",
    "```\n",
    "bcftools stats -s - random_larger_subset.vcf.gz | grep \"^PSC\" -B 1 > hetsTG.txt\n",
    "```\n",
    "\n",
    "If you inspect the file, you will see that there is one line per sample, with several columns containing the statistics. Column 6 is the number of Hets.\n",
    "\n",
    "* Let's have some look at the data (in `R`)!\n",
    "\n",
    "```\n",
    "hets<-read.table(\"hetsTG.txt\", sep=\"\\t\",header=T,comment.char=\"\")\n",
    "hist(hets[,6])\n",
    "```\n",
    "\n",
    "* Is there a relationship between homozygous alternative (1/1) and heterozygous (0/1) sites?\n",
    "\n",
    "```\n",
    "plot(hets[,5],hets[,6])\n",
    "\n",
    "cor.test(hets[,5],hets[,6])\n",
    "```\n",
    "\n",
    "* Now, I would like to stratify this by population groups. For this, we need to load the metadata and do some merging. Unfortunately, when creating this file, the authors messed with empty columns, so we need to tell `R` to ignore this (`fill=T`). Then we merge, taking only the sample ID and heterozygous call columns.\n",
    "\n",
    "\n",
    "```\n",
    "meta<-read.table(\"appladmix/notebooks/2_popgen/data/integrated_call_samples_v3.20130502.ALL.panel\", sep=\"\\t\",header=T,fill=T)\n",
    "\n",
    "head(meta)\n",
    "\n",
    "hetmet<-merge(hets[,c(3,6)],meta,by.x=1,by.y=1)\n",
    "\n",
    "head(hetmet)\n",
    "```\n",
    "\n",
    "* Then, let's make a boxplot stratified by continental population (already adding colour and labels to nice it up):\n",
    "\n",
    "```\n",
    "boxplot(X.6.nHets~super_pop,data=hetmet,\n",
    "    col=c(\"blue\",\"green\",\"orange\",\"yellow\",\"red\"),\n",
    "    xlab=\"Continental population\",ylab=\"Heterozygous sites\")\n",
    "```\n",
    "\n",
    "This very simple statistic does have a biological meaning, hence it is very informative!\n",
    "\n",
    "* A final thing: nicely ordering the data by super-population, and plotting the distribution by more specific populations. This requires a bit of `R`-specific data handling, to get a properly sorted dataframe.\n",
    "\n",
    "```\n",
    "mypops<-c(\"AFR\",\"SAS\",\"EUR\",\"EAS\",\"AMR\")\n",
    "sortedpops<-list()\n",
    "for (npop in mypops) { sortedpops[[npop]]<-unique(hetmet$pop[which(hetmet$super_pop==npop)]) }\n",
    "hetmetnice<-data.frame(nHets=hetmet[,2],pop=hetmet$pop,super_pop=hetmet$super_pop)\n",
    "hetmetnice$pop = factor(hetmetnice$pop, levels=unlist(sortedpops))\n",
    "\n",
    "boxplot(nHets~pop,data=hetmetnice,\n",
    "    xlab=\"Continental population\",ylab=\"Heterozygous sites\",\n",
    "    col=c(rep(\"blue\",length(sortedpops[[1]])),rep(\"green\",length(sortedpops[[2]])),\n",
    "    rep(\"orange\",length(sortedpops[[3]])),rep(\"yellow\",length(sortedpops[[4]])),\n",
    "    rep(\"red\",length(sortedpops[[5]]))))\n",
    "```\n",
    "\n",
    "Now you can see how the heterozygosities differ between super-populations. Boxplots are not an ideal way of presenting data, but convenient in base `R`.\n",
    "One might rather want to get a violin plot. Here, I recommend more specialized `R` packages such as `ggplot2`. I want a violin plot with quantiles at 25\\%, 50\\% and 75\\%. The plot should be customized with `theme()`, starting from a minimal version but with the colours I propose (`scale_fill_manual`), nice axis labels (`axis.text.x`), and some control over the title (`plot.title`), and custom or no labels on the axes (`xlab` and `ylab`). `ggtitle` provides the title content.\n",
    "\n",
    "```\n",
    "library(\"ggplot2\")\n",
    "colvec<-c(\"blue\",\"green\",\"orange\",\"yellow\",\"red\");names(colvec)<-mypops\n",
    "\n",
    "ggplot(hetmetnice) +\n",
    "  theme_minimal() +\n",
    "  geom_violin(mapping=aes(x=pop,y=nHets,fill=super_pop), draw_quantiles = c(0.25, 0.5, 0.75), scale=\"width\" )  + \n",
    "  scale_fill_manual(values=colvec) +\n",
    "  theme(\n",
    "    axis.text.x = element_text(angle = 45, vjust = 1, hjust=1,size=12.5),\n",
    "    plot.title = element_text(face=\"bold\",hjust=0.5,size=15)) +\n",
    "  xlab(\"\")  + ylab(\"Number of heterozygous sites\") + \n",
    "  ggtitle(label=paste(\"1000 Genomes heterozygous sites on chr18\")) \n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "805a75db",
   "metadata": {},
   "source": [
    "## Population clustering\n",
    "\n",
    "The program [ADMIXTURE](https://dalexander.github.io/admixture/) is a widely used tool to estimate proportional ancestries from SNP data. Read more about it [here](https://dx.doi.org/10.1101/gr.094052.109). The funny thing is that his program does not actually detect admixture, but clusters SNPs based on shared occurrence across samples. The reason why they are shared can be various, including admixture, but also simply shared ancestry, bottlenecks, or other events in the history of the populations.\n",
    "\n",
    "In any case, it is often used to understand how individuals/populations relate to each other, or more specifically how many clusters of individuals are there in the population based on their majority ancestry. For example, there might be three population clusters in bonobos, as shown in panel B of this from a [publication](https://doi.org/10.1016/j.cub.2024.09.043) on whole genome data from this species:\n",
    "\n",
    "<img src=\"https://ars.els-cdn.com/content/image/1-s2.0-S0960982224012843-gr1.jpg\">\n",
    "\n",
    "\n",
    "This is what we can also do with the human data at hand.\n",
    "\n",
    "Since this is quite a resource-consuming calculation (scaling with the number of SNPs), we cannot do it on the whole chromosome, so we subsample again, this time aiming for ~50,000 SNVs or 2.5% of the original SNVs (note the change in order of magnitude):\n",
    "\n",
    "```\n",
    "gatk SelectVariants --select-random-fraction 0.025 --variant $TGDATA --output random_subset.vcf.gz\n",
    "```\n",
    "\n",
    "The program does not digest a VCF file, so we have to transform the data into another file format. A suitable format is called the \"plink bed\" format, which we can get with the program [plink](https://www.cog-genomics.org/plink/). This tool and its various data formats are widely used in genomics, but here we simply use it to convert our data the required format, and on the go do some filtering:\n",
    "\n",
    "\n",
    "```\n",
    "plink2 --vcf random_subset.vcf.gz --max-alleles 2 --snps-only --make-bed --maf 0.05 --out random_subset \n",
    "```\n",
    "\n",
    "* Let's inspect what each filtering parameter is actually doing.\n",
    "\n",
    "\n",
    "Now, finally, we run ADMIXTURE, and we assume that there might be two different clusters in the data.\n",
    "\n",
    "```\n",
    "admixture random_subset.bed 2 -j1 > random_2k.out\n",
    "```\n",
    "\n",
    "A good thing is that it does not need to know how many clusters are there, as often we don't know it. So we can run it under the assumption of different numbers of clusters:\n",
    "\n",
    "```\n",
    "admixture random_subset.bed 3 -j1 > random_3k.out\n",
    "admixture random_subset.bed 4 -j1 > random_4k.out\n",
    "admixture random_subset.bed 5 -j1 > random_5k.out\n",
    "```\n",
    "\n",
    "That is indeed pretty easy, just apply a command line tool. We can have a look at the output.\n",
    "\n",
    "But more important would be a visualization like the plot from the publication above.\n",
    "\n",
    "\n",
    "### ADMIXTURE plots\n",
    "Now, we can do this in `R` or `Rstudio`. There is a metadata table in your own directories.\n",
    "\n",
    "```\n",
    "meta<-read.table(\"appladmix/notebooks/2_popgen/data/integrated_call_samples_v3.20130502.ALL.panel\", sep=\"\\t\",header=T,fill=T)\n",
    "```\n",
    "\n",
    "How does the data look?\n",
    "\n",
    "```\n",
    "tabl2=read.table(\"random_subset.2.Q\")\n",
    "tabl5=read.table(\"random_subset.5.Q\")\n",
    "```\n",
    "\n",
    "Ok, so one would need to create something like a stacked barplot with these numbers.\n",
    "\n",
    "```\n",
    "barplot(t(tabl2),beside=F,col=c(\"green\",\"black\"),border=NA)\n",
    "```\n",
    "\n",
    "\n",
    "This is a bit messy, but it works. Making this a nice kind of plot is maybe a bit complicated, and I would use packages to help:\n",
    "\n",
    "```\n",
    "library(\"ggplot2\")\n",
    "library(\"tidyr\")\n",
    "library(\"ggpubr\")\n",
    "```\n",
    "\n",
    "First, I will do some polishing with the `tidyverse` syntax to get things into a nice order. What follows will be a bit complex, but it is indeed just ordering the very same data you saw before. The goal is to separate the different global metapopulations, and display the data in a visually nice way by ordering the individuals within each population by components.\n",
    "\n",
    "Also, this can be done in a loop to save the information for 2 to 5 clusters in an `R` object.\n",
    "\n",
    "\n",
    "```\n",
    "myplots=list()\n",
    "for (i in 2:5) {\n",
    "    tbl=read.table(paste(\"random_subset.\",i,\".Q\",sep=\"\"))\n",
    "    tbl<-cbind(tbl,Label=meta[,3])\n",
    "\n",
    "    plot_data <- tbl %>%\n",
    "        mutate(id = row_number()) %>%\n",
    "        gather('pop', 'prob', V1:paste0(\"V\",i)) %>%\n",
    "        group_by(id) %>%\n",
    "        mutate(likely_assignment = pop[which.max(prob)],\n",
    "               assingment_prob = max(prob)) %>%\n",
    "        arrange(likely_assignment, desc(assingment_prob)) %>%\n",
    "        ungroup() %>%\n",
    "        mutate(id = forcats::fct_inorder(factor(id)))\n",
    "\n",
    "    myplots[[i]]<-\n",
    "    ggplot(plot_data, aes(id, prob, fill = pop)) +\n",
    "           geom_col() +\n",
    "        facet_grid(~Label, scales = 'free', space = 'free') +\n",
    "        theme(\n",
    "          axis.title.x = element_blank(),\n",
    "          axis.text.x = element_blank(),\n",
    "          axis.ticks.x = element_blank()\n",
    "        ) + theme(legend.position = \"none\") + ggtitle(paste0(\"K=\",i))\n",
    "    }\n",
    "```\n",
    "\n",
    "\n",
    "Finally, you can plot everything in one figure (commented out the way to save it as pdf). Again, we use an `R` package to make it look good, basically publication-ready.\n",
    "\n",
    "\n",
    "\n",
    "```\n",
    "#pdf(\"figure_admixture.pdf\", 10, 20)\n",
    "ggarrange(myplots[[2]],myplots[[3]],myplots[[4]],myplots[[5]],ncol=1, nrow=4)\n",
    "#dev.off()\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15d62655",
   "metadata": {},
   "source": [
    "# Principle Component Analysis\n",
    "\n",
    "## [What is a PCA?](https://en.wikipedia.org/wiki/Principal_component_analysis)\n",
    "\n",
    "* A statistical method to reduce dimensionality of the data.\n",
    "* *E.g.* hundreds of individuals with thousands or millions of SNVs\n",
    "* Many SNVs are observed across multiple individuals: co-variance\n",
    "* Co-variance shows how similar individuals are - populations cluster with each other\n",
    "* (unsupervised = the method does not know the labels)\n",
    "\n",
    "\n",
    "### PCA in population genetics\n",
    "\n",
    "It is a widely used summary statistic for genotype data. In many papers, it is shown as an exploratory step - we have sequenced new individuals, where do they fall in worldwide or large-scale diversity? This works for humans and many other species with sufficient amounts of data.\n",
    "\n",
    "Here an example from [humans](https://doi.org/10.1038%2Fnature07331): \n",
    "\n",
    "<img src=\"https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnature07331/MediaObjects/41586_2008_Article_BFnature07331_Fig1_HTML.jpg\">\n",
    "\n",
    "Or, as above, from bonobos, this time focus on panel A:\n",
    "\n",
    "<img src=\"https://ars.els-cdn.com/content/image/1-s2.0-S0960982224012843-gr1.jpg\">\n",
    "\n",
    "\n",
    "\n",
    "## Getting a simple PCA\n",
    "\n",
    "Now, there are many ways to calculate a PCA. Here, we will take one of the recently published softwares, because it is conveniently spitting out the PCA directly from a VCF file!\n",
    "\n",
    "There are some warnings, but closer inspection tells us that it is fine, kind of.\n",
    "\n",
    "```\n",
    "mkdir pca\n",
    "$main/VCF2PCACluster/bin/VCF2PCACluster -i chr21.merged.vcf.gz -o pca/geno_pca\n",
    "```\n",
    "\n",
    "That's it, and that is how a neat tool should work.\n",
    "\n",
    "(SPOILER: most tools don't! Even here you will see some bugs that either appear on this particular computing structure, or accumulate since the program was written.)\n",
    "\n",
    "Now we can inspect the output files as well as the plots this program created.\n",
    "\n",
    "* We see that PC1 is separating Archaics from humans, with Europeans inbetween. Why is that the case?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8b40022",
   "metadata": {},
   "source": [
    "##### More control: doing a simple PCA in R\n",
    "\n",
    "### Data preparation\n",
    "\n",
    "Now, let's do a PCA with a bit more control than this out-of-the-box software. We may re-use the chr18 data from the 1000 Genomes. What we need now is a genotype matrix, or a table of numbers which can be used by a statistical method. This table should also not have missing data, or other stuff.\n",
    "\n",
    "Fortunately, we have already subsampled the large file into a smaller file, which we can re-use again. However, it matters that here we only use SNVs with two states (one REF, one ALT, no further alleles).\n",
    "\n",
    "A neat feature of `bcftools` is that it can also transform `vcf` formatted files into anything you need. In this case, we really just want to have the genotypes as numbers without anything else. `bcftools query` is good for that, since you can define which fields to extract, and how to separate them.\n",
    "\n",
    "* Let's do it in one go!\n",
    "\n",
    "```\n",
    "# in BASH\n",
    "\n",
    "bcftools view -m 2 -M2 -v snps random_subset.vcf.gz | bcftools query -f '[%GT ]\\n' > allgts.txt\n",
    "```\n",
    "\n",
    "* Then we can switch into `R`.\n",
    "\n",
    "\n",
    "### Data preparation in R\n",
    "\n",
    "First, we may need the proper library to calculate this! A possible (and very nice) one is [adegenet](https://adegenet.r-forge.r-project.org/). `R` has many packages, and often they are not pre-installed on a computer/structure. If that is the case, you may need to install a package with:\n",
    "\n",
    "```\n",
    "library(adegenet)\n",
    "```\n",
    "\n",
    "Now, read the SNPS and have a look!\n",
    "\n",
    "```\n",
    "snps<-read.table(\"allgts.txt\", sep=\" \",header=F,comment.char=\"\")\n",
    "head(snps)\n",
    "```\n",
    "\n",
    "* Note that the separator is now a space, because that is how we defined it in the previous step. However, this was adding a last space to the end of the file, and R believes there is a (empty) column.\n",
    "\n",
    "```\n",
    "snps<-snps[,-2505]\n",
    "\n",
    "table(snps[,1])\n",
    "```\n",
    "\n",
    "Now, these are not really numbers, but genotypes...\n",
    "\n",
    "* A very easy way to turn them into single numbers would be using the `ifelse` statement like this:\n",
    "\n",
    "```\n",
    "snps[,1]<-ifelse(snps[,1]==\"0|0\",0,snps[,1])\n",
    "snps[,1]<-ifelse(snps[,1]==\"1|1\",2,snps[,1])\n",
    "snps[,1]<-ifelse(snps[,1]%in%c(\"0|1\",\"1|0\"),1,snps[,1])\n",
    "table(snps[,1])\n",
    "```\n",
    "\n",
    "* However, this is not numeric, and mathematical R functions are picky about this:\n",
    "\n",
    "```\n",
    "is.numeric(snps[,1])\n",
    "snps[,1]<-as.numeric(snps[,1])\n",
    "is.numeric(snps[,1])\n",
    "```\n",
    "\n",
    "This looks better, but you certainly don't want to type in 2503 more times the same things, and you would very likely make mistakes on the way.\n",
    "\n",
    "* Let's use a `for` loop and nested `ifelse` statements to make it easier! As a side-effect, the whole column will be numeric as well!\n",
    "\n",
    "```\n",
    "for (j in (2:2504)) { snps[,j]<-ifelse(snps[,j]==\"0|0\",0,ifelse(snps[,j]==\"1|1\",2,ifelse(snps[,j]%in%c(\"0|1\",\"1|0\"),1,NA))); print(j) }\n",
    "\n",
    "\n",
    "table(snps[,3])\n",
    "table(snps[,666])\n",
    "is.numeric(snps[,2024])\n",
    "```\n",
    "\n",
    "I would note that at this stage it is ok to do things that way. But next level would be to write a `function` to make it more efficient...\n",
    "\n",
    "If you feel that coming up with efficient solutions is complicated, that is fine - it is a matter of practice and experience. We need to learn things in order to do them.\n",
    "\n",
    "\n",
    "### Calculating the PCA\n",
    "\n",
    "The method we want to use is called `dudi.pca`. It needs the individuals in the rows, and the positions in the columns, that is, a transposed table to be done with `t()`. As you see, nested functions are very common in `R`, do combine things on the fly without intermediate objects. \n",
    "\n",
    "* Let's to this:\n",
    "\n",
    "```\n",
    "pca_object <- dudi.pca(t(snps),nf=20,scannf=F)\n",
    "\n",
    "tb<-list()\n",
    "for (j in (1:ncol(snps)))  { tb[[j]]<-table(snps[,j]) }\n",
    "\n",
    "pca_object\n",
    "summary(pca_object)\n",
    "```\n",
    "\n",
    "This works so far! We have created a higher-level `R` object with lots of information. Important are the coordinates for each individual, which are in the sub-object called `li`, which we can access with the `$` sign.\n",
    "\n",
    "### Plotting the PCA\n",
    "\n",
    "* Let's plot this!\n",
    "\n",
    "```\n",
    "#pdf(\"PCA_1000g.pdf\",12,12)\n",
    "plot(pca_object$li[,1],pca_object$li[,2])\n",
    "#dev.off()\n",
    "```\n",
    "\n",
    "Ok, this is just some distribution of dots... It is nicer to color it. For this, we need the metadata again. In `R`, we can then create a matching colour vector:\n",
    "\n",
    "```\n",
    "meta<-read.table(\"appladmix/notebooks/2_popgen/data/integrated_call_samples_v3.20130502.ALL.panel\", sep=\"\\t\",header=T,fill=T)\n",
    "mypops<-c(\"AFR\",\"SAS\",\"EUR\",\"EAS\",\"AMR\")\n",
    "cols=c(\"blue\",\"green\",\"orange\",\"yellow\",\"red\")\n",
    "\n",
    "colvector<-meta$super_pop\n",
    "for (pop in (1:length(mypops))) { colvector<-gsub(mypops[pop],cols[pop],colvector) }\n",
    "```\n",
    "\n",
    "* Let's plot it!\n",
    "\n",
    "```\n",
    "#pdf(\"PCA_1000g_col.pdf\",12,12)\n",
    "plot(pca_object$li[,1],pca_object$li[,2],col=colvector)\n",
    "legend(\"top\",legend=mypops, fill=cols,bty=\"n\" )\n",
    "#dev.off()\n",
    "```\n",
    "\n",
    "This is actually already quite good! We can do some more polishing to make it look nicer, and add the information how much of the variation is explained by these first two PCs (usually that is helpful to judge how informative this is in the context of population diversity).\n",
    "\n",
    "* Let's add nice axis labels (`xlab` and `ylab`), and points instead of empty circles (with `pch`):\n",
    "\n",
    "\n",
    "```\n",
    "percent_variation=round(pca_object$eig/sum(pca_object$eig)*100,2)\n",
    "\n",
    "#pdf(\"PCA_1000g_nice.pdf\",12,12)\n",
    "plot(pca_object$li[,1],pca_object$li[,2],col=colvector,pch=16,\n",
    "    xlab=paste(\"PC1\",\" (\",percent_variation[1],\"%)\",sep=\"\"),\n",
    "    ylab=paste(\"PC\",2,\" (\",percent_variation[2],\"%)\",sep=\"\"),\n",
    "    main=paste(\"PCA of 1000G, chr18\"))\n",
    "legend(\"topright\",legend=mypops, fill=cols,bty=\"n\" )\n",
    "#dev.off()\n",
    "```\n",
    "\n",
    "Now, this looks good, and you can gain some insight into the relationships of these populations!\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fedd8f1",
   "metadata": {},
   "source": [
    "# Now, let's go for Challenge 1!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "716198f5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
